### PR TITLE
Support development requirements in Vagrant and Docker devstacks

### PR DESCRIPTION
**Pull request by andy-armstrong**
Configuration Pull Request

@clintonb @doctoryes is this roughly what you were thinking? I'll look into building these tomorrow if so.

---

Make sure that the following steps are done before merging:

  - [ ] A DevOps team member has approved the PR.
  - [ ] Are you adding any new default values that need to be overridden when this change goes live? If so:
    - [ ] Update the appropriate internal repo (be sure to update for all our environments)
    - [ ] If you are updating a secure value rather than an internal one, file a DEVOPS ticket with details.
    - [ ] Add an entry to the CHANGELOG.
  - [ ] Have you performed the proper testing specified on the [Ops Ansible Testing Checklist](https://openedx.atlassian.net/wiki/display/EdxOps/Ops+Ansible+Testing+Checklist)?

### CODE REVIEW

File: docker/build/edxapp/devstack.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+---
```
Comment (clintonb) [inline docker/build/edxapp/devstack.yml]: Can we symlink this?

**Review by clintonb — COMMENTED** [review]
The failure is due to Docker not knowing about the new file. That is because it is ignored: https://github.com/edx/configuration/blob/5458bbd9a317288ed29774f95716baeeb8ae4322/.dockerignore. Adding an entry to `.dockerignore` should fix the failure. 

More details: https://docs.docker.com/engine/reference/builder/#dockerignore-file.

File: docker/build/edxapp/devstack.yml (hunk 1 above)
Comment (andy-armstrong) [reply docker/build/edxapp/devstack.yml]: Good question. Is that a reasonable thing to do? I know that there are issues with Windows and symlinks.

Comment (clintonb) [reply docker/build/edxapp/devstack.yml]: I believe we already use a symlink for the playbooks in the docker directory. @edx/devops can chime in regarding whether this behavior should be replicated.

File: docker/build/edxapp/devstack.yml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+# of requirements should not be used in production.
+pre_requirements_file:    "{{ edxapp_code_dir }}/requirements/edx/pre.txt"
+github_requirements_file: "{{ edxapp_code_dir }}/requirements/edx/github.txt"
+custom_requirements_file:   "{{ edxapp_code_dir }}/requirements/edx/custom.txt"
+local_requirements_file:  "{{ edxapp_code_dir }}/requirements/edx/local.txt"
+base_requirements_file:   "{{ edxapp_code_dir }}/requirements/edx/base.txt"
+post_requirements_file:   "{{ edxapp_code_dir }}/requirements/edx/post.txt"
+paver_requirements_file:   "{{ edxapp_code_dir }}/requirements/edx/paver.txt"
+development_requirements_file:   "{{ edxapp_code_dir }}/requirements/edx/development.txt"
+testing_requirements_file:   "{{ edxapp_code_dir }}/requirements/edx/testing.txt"
```
Comment (doctoryes) [inline docker/build/edxapp/devstack.yml]: It looks like you can just use the playbook defaults for most of these defs:
https://github.com/edx/configuration/blob/master/playbooks/roles/edxapp/defaults/main.yml#L1205-L1212
You'd only need to define development_requirements_file, testing_requirements_file, and edxapp_requirements_files.

File: docker/build/edxapp/devstack.yml (hunk 1 above)
Comment (doctoryes) [reply docker/build/edxapp/devstack.yml]: There are existing symlinks in the configuration repo:
```
NE55-SIX-EIGHTEEN:configuration jeskew$ find . -type d -exec ls -la {} \; -print | egrep ^l
lrwxr-xr-x  1 jeskew staff   29 Mar 15  2016 analytics_api_tmp.yml -> ../../plays/analytics_api.yml
lrwxr-xr-x  1 jeskew staff  24 Mar 15  2016 library -> ../../playbooks/library/
lrwxr-xr-x  1 jeskew staff  22 Mar 15  2016 roles -> ../../playbooks/roles/
lrwxr-xr-x  1 jeskew staff    8 May 26  2016 roles -> ../roles
lrwxr-xr-x  1 jeskew staff   19 Mar 15  2016 callback_plugins -> ../callback_plugins
lrwxr-xr-x  1 jeskew staff    9 Mar 15  2016 ec2.py -> ../ec2.py
lrwxr-xr-x  1 jeskew staff   13 Mar 15  2016 group_vars -> ../group_vars
lrwxr-xr-x  1 jeskew staff   16 Mar 15  2016 inventory.ini -> ../inventory.ini
lrwxr-xr-x  1 jeskew staff   10 Mar 15  2016 library -> ../library
lrwxr-xr-x  1 jeskew staff    8 Mar 15  2016 roles -> ../roles
lrwxr-xr-x   1 jeskew staff   10 Nov  8  2016 mongo -> mongo_2_6/
lrwxr-xr-x 1 jeskew staff   30 Mar 15  2016 ansible.cfg -> ../../../playbooks/ansible.cfg
lrwxr-xr-x 1 jeskew staff   30 Mar 15  2016 ansible.cfg -> ../../../playbooks/ansible.cfg
lrwxr-xr-x 1 jeskew staff   30 Mar 15  2016 ansible.cfg -> ../../../playbooks/ansible.cfg
lrwxr-xr-x 1 jeskew staff   30 Mar 15  2016 ansible.cfg -> ../../../playbooks/ansible.cfg
```
So the question is whether it's something we want to continue doing or not.

**Comment by clintonb** [comment]
I updated `.dockerignore`.

**Comment by clintonb** [comment]
@doctoryes' commit fixed the broken build. I have squashed the commits, and rebased to resolve the merge conflict. This should be good to go once the build passes for the final commit. #teamwork

**Review by clintonb — APPROVED** [review]

**Comment by doctoryes** [comment]
@clintonb The symlink issue hasn't been cleared up yet. I'm in favor of having one file that gets symlinked. I'm willing to make that change as well.

**Comment by clintonb** [comment]
@doctoryes Yeah. Go for it. I'll leave it to you to merge.

File: playbooks/roles/edxapp/vars/devstack.yml [hunk 3]
```diff
@@ -1,14 +1,8 @@
...
-# defaults is the presence of the "testing.txt" file, which is used to install
-# packages not used in production
-pre_requirements_file:    "{{ edxapp_code_dir }}/requirements/edx/pre.txt"
-github_requirements_file: "{{ edxapp_code_dir }}/requirements/edx/github.txt"
-custom_requirements_file:   "{{ edxapp_code_dir }}/requirements/edx/custom.txt"
-local_requirements_file:  "{{ edxapp_code_dir }}/requirements/edx/local.txt"
-base_requirements_file:   "{{ edxapp_code_dir }}/requirements/edx/base.txt"
-post_requirements_file:   "{{ edxapp_code_dir }}/requirements/edx/post.txt"
-paver_requirements_file:   "{{ edxapp_code_dir }}/requirements/edx/paver.txt"
+# The only difference between these requirements and the role defaults is the
```
Comment (feanil) [inline playbooks/roles/edxapp/vars/devstack.yml:2]: Why is this file needed for non docker builds?  Is vagrant going to start using it?

Comment (doctoryes) [reply playbooks/roles/edxapp/vars/devstack.yml:2]: This PR sprang from the desire to have isort included in the devstack build - for both vagrant and docker. The development.txt file was added long ago by myself - but never integrated into the devstack build. We decided putting the isort dependency there would be the best place - installing it in devstacks but not production AMIs.

The idea is that now we could move more things that shouldn't be in production builds out of base.txt and into development.txt.

**Comment by nedbat** [comment]
I don't understand why development.txt and testing.txt are not installed by paver.  Our multi-headed requirements installation process continues to baffle us :)

**Comment by doctoryes** [comment]
@nedbat I'm baffled as well. Seems like both files should be installed in prereqs.py, especially since paver is only used in devstacks?
Then there's the whole problem of bok-choy, sphinx, nose, etc. being in base.txt and therefore being installed in all production AMIs...

**Comment by nedbat** [comment]
Once we have these files in place, we can move things like nose, sphinx, and coverage (but not mock!!) out of production.

**Comment by doctoryes** [comment]
@jibsheet No rush - low priority - but this PR still awaits a devops approval.

**Comment by jibsheet** [comment]
@doctoryes Ah, yes, @feanil had swung by with a reasonable question and I wanted to ensure he was content with the answer.

File: playbooks/roles/edxapp/vars/devstack.yml (hunk 3 above)
Comment (feanil) [reply playbooks/roles/edxapp/vars/devstack.yml:2]: Makes sense. :+1:

**Review by feanil — APPROVED** [review]
